### ISSUE TITLE
Unable to upload files while masquerading

### ISSUE THREAD

**Issue by va7map** [instructure/canvas-lms#587]
We started hearing reports from teachers saying they are not able to upload assignment submissions while acting as a test student (a.k.a. Student View).

Here are the steps to reproduce:

1. As a teacher, create an assignment that takes file uploads
2. Go to course settings and begin Student View
3. As the student, navigate to the assignment, and click Submit Assignment in the right sidebar
4. Select a file to upload, then click the Submit Assignment button

The button text will change to "Submitting..." and stay frozen. If you access another other page in Canvas (or reload the same page), you will no longer be in Student View. And if you inspect the response, you will see that Canvas returned a *406 Not Acceptable* error to the `POST` request sent to `/users/:user_id/files.text`.

In step 2, instead of using Student View, you can masquerade as any student enrolled in the course. The result should be the same.

---

Our installation uses local file storage, and we have Redis and CAS enabled. After some debugging, it looks like that's the exact setup required for this issue to occur.

At some point during the file upload request, `AuthenticationMethods#load_user` is called. If CAS authentication is enabled, it checks if the CAS ticket stored in the session has expired by calling [`Pseudonym#cas_ticket_expired?`](https://github.com/instructure/canvas-lms/blob/a95b729ac6a4ec112a8916abca50129aaf54f455/app/models/pseudonym.rb#L522-L529):

```ruby
def cas_ticket_expired?(ticket)
  return unless Canvas.redis_enabled?
  redis_key = Pseudonym.cas_ticket_key(ticket)

  # Refresh the ttl on the cas ticket before we check its state.
  Canvas.redis.expire(redis_key, CAS_TICKET_TTL)
  Canvas.redis.get(redis_key) != global_id.to_s
end
```

Regardless of whether you're masquerading, the CAS ticket stored in the session is the same one you obtained when you first logged in, as yourself. This means the value cached in Redis (LHS of the `!=` operator) is your own pseudonym's `global_id`. However, while masquerading as another user, the `global_id` being checking against (RHS) will be that of the other user's pseudonym, which won't match.

In other words, while masquerading, `Pseudonym#cas_ticket_expired?` will always return `true`, given that Redis is enabled.

**Comment by ccutrer** [comment]
Are you sure about your conclusion? `Pseudonym#cas_ticket_expired?` is only called from `AuthenticationMethods#load_user`, and is called on `@current_pseudonym` _before_ masquerading is applied (later on in the same method). If you were correct, simply masquerading at all while using CAS would immediately log you out, and that seems like an issue that would immediately be reported as super high priority.

**Comment by va7map** [comment]
You're right. Masquerading is applied *after* `Pseudonym#cas_ticket_expired?`. Otherwise, we'd always be logged out immediately.

The issue only occurs when uploading files. So I dug deeper, and found this in the `FilesController`:

````ruby
prepend_around_filter :load_pseudonym_from_policy, :only => :create
````

The `POST` request sent to `/users/:user_id/files.text` (i.e. `Files#create`) has a `Policy` parameter, and it contains the `pseudonym_id` of the user being masqueraded as (the "impostee"). This causes `@policy_pseudonym_id` to be set by `AuthenticationMethods#load_pseudonym_from_policy`, which in turn causes `@current_pseudonym` to become the impostee's pseudonym, instead of mine.

I believe this is why file upload fails while masquerading.

**Comment by ccutrer** [comment]
Ah, nice sleuthing. Does the commit that was just mentioned fix it for you?

**Comment by va7map** [comment]
Yes, it did!

**Comment by simonista** [comment]
@va7map, I just wanted to add that we really appreciated the detailed bug report and investigation you did here!

**Comment by va7map** [comment]
You're welcome, @simonista! Thanks to the nudge from @ccutrer, I made the connection and totally enjoyed the "ah ha!" moment.
